File: task3-unknown-language/Elixir_notes_app/lib/notes_app/handler/note_handler.ex
defmodule NotesApp.Handler.NoteHandler do
  @moduledoc """
  HTTP обработчики для endpoints управления заметками.
  Обрабатывает JSON запросы и ответы, валидацию и ошибки.
  """

  import Plug.Conn
  alias NotesApp.Service.NoteService

  @doc """
  GET /notes - получить все заметки
  """
  def list_notes(conn) do
    notes = NoteService.list_notes()
    
    conn
    |> put_resp_content_type("application/json")
    |> send_resp(200, Jason.encode!(%{notes: notes}))
  end

  @doc """
  POST /notes - создать новую заметку
  """
  def create_note(conn) do
    case parse_json_body(conn) do
      {:ok, body} ->
        case NoteService.create_note(body) do
          {:ok, note} ->
            conn
            |> put_resp_content_type("application/json")
            |> send_resp(201, Jason.encode!(note))
          
          {:error, reason} ->
            send_error(conn, 400, reason)
        end
      
      {:error, reason} ->
        send_error(conn, 400, reason)
    end
  end

  @doc """
  PUT /notes/:id - обновить заметку
  """
  def update_note(conn, id) do
    case parse_json_body(conn) do
      {:ok, body} ->
        case NoteService.update_note(id, body) do
          {:ok, note} ->
            conn
            |> put_resp_content_type("application/json")
            |> send_resp(200, Jason.encode!(note))
          
          {:error, :not_found} ->
            send_error(conn, 404, "Note not found")
          
          {:error, :invalid_id} ->
            send_error(conn, 400, "Invalid note ID")
          
          {:error, reason} ->
            send_error(conn, 400, reason)
        end
      
      {:error, reason} ->
        send_error(conn, 400, reason)
    end
  end

  @doc """
  DELETE /notes/:id - удалить заметку
  """
  def delete_note(conn, id) do
    case NoteService.delete_note(id) do
      {:ok, _note} ->
        conn
        |> send_resp(204, "")
      
      {:error, :not_found} ->
        send_error(conn, 404, "Note not found")
      
      {:error, :invalid_id} ->
        send_error(conn, 400, "Invalid note ID")
      
      {:error, reason} ->
        send_error(conn, 500, reason)
    end
  end

  # Приватные функции

  defp parse_json_body(conn) do
    case read_body(conn) do
      {:ok, body, _conn} ->
        case Jason.decode(body) do
          {:ok, json} -> {:ok, json}
          {:error, _} -> {:error, "Invalid JSON"}
        end
      
      {:error, _} ->
        {:error, "Failed to read request body"}
    end
  end

  defp send_error(conn, status, message) do
    error_response = %{error: message}
    
    conn
    |> put_resp_content_type("application/json")
    |> send_resp(status, Jason.encode!(error_response))
  end
end

File: task3-unknown-language/Go_notes_app/internal/handler/note_handler.go
package handler

import (
	"encoding/json"
	"errors"
	"net/http"
	"noteapp/internal/model"
	"noteapp/internal/service"
	"strconv"

	"github.com/go-chi/chi/v5"
)

// NoteHandler обработчик HTTP запросов для заметок
type NoteHandler struct {
	service service.NoteService
}

// NewNoteHandler создает новый экземпляр обработчика заметок
func NewNoteHandler(service service.NoteService) *NoteHandler {
	return &NoteHandler{service: service}
}

// GetAllNotes обрабатывает GET /notes - получение всех заметок
func (h *NoteHandler) GetAllNotes(w http.ResponseWriter, r *http.Request) {
	notes, err := h.service.GetAllNotes()
	if err != nil {
		h.writeErrorResponse(w, http.StatusInternalServerError, "Failed to get notes")
		return
	}

	h.writeJSONResponse(w, http.StatusOK, notes)
}

// CreateNote обрабатывает POST /notes - создание новой заметки
func (h *NoteHandler) CreateNote(w http.ResponseWriter, r *http.Request) {
	var req model.CreateNoteRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.writeErrorResponse(w, http.StatusBadRequest, "Invalid JSON format")
		return
	}

	note, err := h.service.CreateNote(&req)
	if err != nil {
		if errors.Is(err, model.ErrEmptyText) || errors.Is(err, model.ErrTextTooLong) {
			h.writeErrorResponse(w, http.StatusBadRequest, err.Error())
			return
		}
		h.writeErrorResponse(w, http.StatusInternalServerError, "Failed to create note")
		return
	}

	h.writeJSONResponse(w, http.StatusCreated, note)
}

// UpdateNote обрабатывает PUT /notes/{id} - обновление заметки
func (h *NoteHandler) UpdateNote(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	id, err := strconv.Atoi(idStr)
	if err != nil {
		h.writeErrorResponse(w, http.StatusBadRequest, "Invalid note ID")
		return
	}

	var req model.UpdateNoteRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.writeErrorResponse(w, http.StatusBadRequest, "Invalid JSON format")
		return
	}

	note, err := h.service.UpdateNote(id, &req)
	if err != nil {
		if errors.Is(err, model.ErrNoteNotFound) {
			h.writeErrorResponse(w, http.StatusNotFound, "Note not found")
			return
		}
		if errors.Is(err, model.ErrEmptyText) || errors.Is(err, model.ErrTextTooLong) {
			h.writeErrorResponse(w, http.StatusBadRequest, err.Error())
			return
		}
		h.writeErrorResponse(w, http.StatusInternalServerError, "Failed to update note")
		return
	}

	h.writeJSONResponse(w, http.StatusOK, note)
}

// DeleteNote обрабатывает DELETE /notes/{id} - удаление заметки
func (h *NoteHandler) DeleteNote(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	id, err := strconv.Atoi(idStr)
	if err != nil {
		h.writeErrorResponse(w, http.StatusBadRequest, "Invalid note ID")
		return
	}

	err = h.service.DeleteNote(id)
	if err != nil {
		if errors.Is(err, model.ErrNoteNotFound) {
			h.writeErrorResponse(w, http.StatusNotFound, "Note not found")
			return
		}
		h.writeErrorResponse(w, http.StatusInternalServerError, "Failed to delete note")
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

// ErrorResponse представляет структуру ответа с ошибкой
type ErrorResponse struct {
	Error string `json:"error"`
}

// writeJSONResponse записывает JSON ответ
func (h *NoteHandler) writeJSONResponse(w http.ResponseWriter, statusCode int, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(statusCode)
	json.NewEncoder(w).Encode(data)
}

// writeErrorResponse записывает ответ с ошибкой
func (h *NoteHandler) writeErrorResponse(w http.ResponseWriter, statusCode int, message string) {
	h.writeJSONResponse(w, statusCode, ErrorResponse{Error: message})
}